src/coordinator/ts/m3db/convert.go

```diff
@@ -0,0 +1,168 @@
+func ConvertM3DBSeriesIterators(iterators encoding.SeriesIterators, iterAlloc encoding.ReaderIteratorAllocate) ([]SeriesBlocks, error) {
+	defer iterators.Close()
+	multiSeriesBlocks := make([]SeriesBlocks, iterators.Len())
+
+	for i, seriesIterator := range iterators.Iters() {
+		blockReplicas, err := blockReplicasFromSeriesIterator(seriesIterator, iterAlloc)
+		if err != nil {
+			return nil, err
+		}
+
+		series, err := seriesBlocksFromBlockReplicas(blockReplicas, seriesIterator)
+		if err != nil {
+			return nil, err
+		}
+
+		multiSeriesBlocks[i] = series
+	}
+
+	return multiSeriesBlocks, nil
+}
+
+func blockReplicasFromSeriesIterator(seriesIterator encoding.SeriesIterator, iterAlloc encoding.ReaderIteratorAllocate) ([]BlockReplica, error) {
+	blockReplicas := make([]BlockReplica, 0, blockReplicaLen)
+	for _, replica := range seriesIterator.Replicas() {
+		perBlockSliceReaders := replica.Readers()
+		for next := true; next; next = perBlockSliceReaders.Next() {
+			l, start, bs := perBlockSliceReaders.CurrentReaders()
+			readers := make([]xio.SegmentReader, l)
+			for i := 0; i < l; i++ {
+				reader := perBlockSliceReaders.CurrentReaderAt(i)
+				// NB(braskin): important to clone the reader as we need its position reset before
+				// we use the contents of it again
+				clonedReader, err := reader.Clone()
+				if err != nil {
+					return nil, err
+				}
+				readers[i] = clonedReader
+			}
+			// todo(braskin): pooling
+			iter := encoding.NewMultiReaderIterator(iterAlloc, nil)
+			iter.Reset(readers, start, bs)
+
+			inserted := false
+			for i := range blockReplicas {
+				if blockReplicas[i].Start.Equal(start) {
+					inserted = true
+					blockReplicas[i].Replicas = append(blockReplicas[i].Replicas, iter)
+					break
+				}
+			}
+			if !inserted {
+				blockReplicas = append(blockReplicas, BlockReplica{
+					Start:     start,
+					BlockSize: bs,
+					Replicas:  []encoding.MultiReaderIterator{iter},
+				})
+			}
+		}
+	}
+
+	return blockReplicas, nil
+}
+
+func seriesBlocksFromBlockReplicas(blockReplicas []BlockReplica, seriesIterator encoding.SeriesIterator) (SeriesBlocks, error) {
+	var (
+		// todo(braskin): use ident pool
+		clonedID        = ident.StringID(seriesIterator.ID().String())
+		clonedNamespace = ident.StringID(seriesIterator.Namespace().String())
+	)
+
+	clonedTags, err := cloneTagIterator(seriesIterator.Tags())
+	if err != nil {
+		return SeriesBlocks{}, err
+	}
+
+	series := SeriesBlocks{
+		ID:        clonedID,
+		Namespace: clonedNamespace,
+		Tags:      clonedTags,
+	}
+
+	for _, block := range blockReplicas {
+		filterValuesStart := seriesIterator.Start()
+		if block.Start.After(filterValuesStart) {
+			filterValuesStart = block.Start
+		}
+
+		end := block.Start.Add(block.BlockSize)
+
+		filterValuesEnd := seriesIterator.End()
+		if end.Before(filterValuesEnd) {
+			filterValuesEnd = end
+		}
+
+		// todo(braskin): pooling
+		valuesIter := encoding.NewSeriesIterator(clonedID, clonedNamespace,
```

Comment:
We *may* run into some difficulty down the line if we're using the id/namespace/tags form the series directly rather than from the SeriesBlock, since `.Close()` on any iterator will free the bytes used by the rest as well, which is unexpected behaviour

If we only use the values from the SeriesBlock itself we should be safe, but if there's a chance we'd be using the seriesIterators directly, it's safer to clone each of these.

Either way should put a comment noting that we should be really careful using this if anything is shared